source/extensions/filters/udp/dns_filter/dns_filter.cc

```diff
@@ -76,21 +109,149 @@ void DnsFilter::onData(Network::UdpRecvData& client_request) {
     return;
   }
 
-  // TODO(abaptiste): Resolve the requested name
+  // Resolve the requested name
+  const auto response = getResponseForQuery(query_context);
 
-  // Send an answer to the client
+  // We were not able to satisfy the request locally. Return an empty response to the client
+  if (response == DnsLookupResponseCode::Failure) {
+    sendDnsResponse(std::move(query_context));
+    return;
+  }
+
+  // TODO(abaptiste): external resolution
+
+  // We have an answer. Send it to the client
   sendDnsResponse(std::move(query_context));
 }
 
 void DnsFilter::sendDnsResponse(DnsQueryContextPtr query_context) {
   Buffer::OwnedImpl response;
-  // TODO(abaptiste): serialize and return a response to the client
 
+  // Serializes the generated response to the parsed query from the client. If there is a
+  // parsing error or the incoming query is invalid, we will still generate a valid DNS response
+  message_parser_.buildResponseBuffer(query_context, response);
   Network::UdpSendData response_data{query_context->local_->ip(), *(query_context->peer_),
                                      response};
   listener_.send(response_data);
 }
 
+DnsLookupResponseCode DnsFilter::getResponseForQuery(DnsQueryContextPtr& context) {
+  /* It appears to be a rare case where we would have more than one query in a single request.
+   * It is allowed by the protocol but not widely supported:
+   *
+   * See: https://www.ietf.org/rfc/rfc1035.txt
+   *
+   * The question section is used to carry the "question" in most queries,
+   * i.e., the parameters that define what is being asked. The section
+   * contains QDCOUNT (usually 1) entries.
+   */
+  for (const auto& query : context->queries_) {
+    // Try to resolve the query locally. If forwarding the query externally is disabled we will
+    // always attempt to resolve with the configured domains
+    if (isKnownDomain(query->name_) || !config_->forwardQueries()) {
```

Comment:
so if a request is not for a known domain, and forward queries is false, we are trying to resolve via the cluster name? I didn't fully understand why.

Reply:
No, we would no try to resolve it via a cluster.   The domain has to be added to the known suffixes list in order for the filter to attempt resolution.

Looking at clusters is just another progression in the list if the domain is known.